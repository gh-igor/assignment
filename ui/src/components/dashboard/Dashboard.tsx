import React from "react";
import { Post } from "../../dto/Post";

function groupBy<T>(array: T[], key: keyof T, valueResolver?: (value: string) => string) {
    return array.reduce<Record<string, T[]>>((acc, item) => {
        let groupingValue = item[key] as string;
        if (valueResolver) {
            groupingValue = valueResolver(groupingValue);
        }
        acc[groupingValue] = [...(acc[groupingValue] || []), item];
        return acc;
    }, {});
}

type UserStatistics = {
    postsTotalCount: number;
    postCharactersMedian: number;
    postsPerMonth: Record<string, number>;
    longestPost: string;
};

type Props = {
    posts: Post[];
};

const Dashboard = ({ posts }: Props) => {
    const postsGroupedByUser = groupBy<Post>(posts, "from_id");
    const statisticsPerUser: Record<string, UserStatistics> = {};
    for (const userId in postsGroupedByUser) {
        postsGroupedByUser[userId].sort((a, b) => a.message.length - b.message.length);
        const postsLength = postsGroupedByUser[userId].length;
        const postsHalfLength = Math.floor(postsGroupedByUser[userId].length / 2);
        const postsGroupedByMonth = groupBy<Post>(postsGroupedByUser[userId], "created_time", (datetime) =>
            new Date(datetime).toLocaleDateString("en-US", { month: "long" }));
        const postsPerMonth: UserStatistics["postsPerMonth"] = {};
        for (const month in postsGroupedByMonth) {
            postsPerMonth[month] = postsGroupedByMonth[month].length;
        }

        statisticsPerUser[userId] = {
            postsTotalCount: postsLength,
            postCharactersMedian: postsGroupedByUser[userId][postsHalfLength].message.length,
            postsPerMonth,
            longestPost: postsGroupedByUser[userId][postsLength - 1].message,
        };
    }

    return (
        <ul>
            {Object.keys(statisticsPerUser).map(userId => (
                <li key={userId}>
                    <strong>{userId}</strong>
                    <div>postsTotalCount: <strong>{statisticsPerUser[userId].postsTotalCount}</strong></div>
                    <div>postCharactersMedian: <strong>{statisticsPerUser[userId].postCharactersMedian}</strong></div>
                    <ul>postsPerMonth: {Object.keys(statisticsPerUser[userId].postsPerMonth).map(month => (
                        <li key={month}>
                            <div>{month}: <strong>{statisticsPerUser[userId].postsPerMonth[month]}</strong></div>
                        </li>
                    ))}</ul>
                    <div>longestPost: {statisticsPerUser[userId].longestPost}</div>
                </li>
            ))}
        </ul>
    );
};

export default Dashboard;
