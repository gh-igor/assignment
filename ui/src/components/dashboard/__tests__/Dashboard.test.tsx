import React from "react";
import renderer from "react-test-renderer";
import Dashboard from "../Dashboard";

const posts = [
    {
        "id": "post63c92774936de_6efc5692",
        "from_name": "Isidro Schuett",
        "from_id": "user_16",
        "message": "dead press adoption inn thin thumb horseshoe uncle seed inquiry heal bay climate freedom slap thoughtful queen company acquisition abuse squash pattern peasant condition export assessment raid helmet linen water flat dirty fling chord correction ethics murder reconcile virgin letter swallow bullet norm experiment damage dirty gravel manage refuse refuse tap velvet relax provide line speculate bend traction tired transport charter delay pick notice lion coincide initiative modernize option bake risk popular delete roar beer agreement debt tumour lie stress key glow lot reconcile patch fountain competition allocation eject golf south belly corn lion golf trench mug carry relation kinship",
        "type": "status",
        "created_time": "2023-01-19T05:23:14+00:00"
    },
    {
        "id": "post63c92774936f7_0e182cd4",
        "from_name": "Quyen Pellegrini",
        "from_id": "user_19",
        "message": "mess money hate conductor company duck indulge fly seller neighborhood rush message flush empire upset due climate opposition virgin shy dimension tolerate horror wagon run accumulation huge complex mother resource core era cash",
        "type": "status",
        "created_time": "2023-01-19T01:23:00+00:00"
    },
    {
        "id": "post63c92774936fe_8f164a11",
        "from_name": "Lael Vassel",
        "from_id": "user_0",
        "message": "leaflet execute systematic corn avenue porter attention kit horseshoe trench constitution canvas representative fill nest racism occasion crystal resign lake umbrella revolution tile snack cottage carve mess accountant tired faithful aluminium competition follow variant button basket advice withdrawal bury",
        "type": "status",
        "created_time": "2023-01-18T19:43:34+00:00"
    },
];

describe("renders Dashboard", () => {
    test("matches snapshot", () => {
        const tree = renderer.create(<Dashboard posts={posts} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
