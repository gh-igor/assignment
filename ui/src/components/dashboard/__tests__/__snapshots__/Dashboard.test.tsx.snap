// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Dashboard matches snapshot 1`] = `
<ul>
  <li>
    <strong>
      user_16
    </strong>
    <div>
      postsTotalCount: 
      <strong>
        1
      </strong>
    </div>
    <div>
      postCharactersMedian: 
      <strong>
        694
      </strong>
    </div>
    <ul>
      postsPerMonth: 
      <li>
        <div>
          January
          : 
          <strong>
            1
          </strong>
        </div>
      </li>
    </ul>
    <div>
      longestPost: 
      dead press adoption inn thin thumb horseshoe uncle seed inquiry heal bay climate freedom slap thoughtful queen company acquisition abuse squash pattern peasant condition export assessment raid helmet linen water flat dirty fling chord correction ethics murder reconcile virgin letter swallow bullet norm experiment damage dirty gravel manage refuse refuse tap velvet relax provide line speculate bend traction tired transport charter delay pick notice lion coincide initiative modernize option bake risk popular delete roar beer agreement debt tumour lie stress key glow lot reconcile patch fountain competition allocation eject golf south belly corn lion golf trench mug carry relation kinship
    </div>
  </li>
  <li>
    <strong>
      user_19
    </strong>
    <div>
      postsTotalCount: 
      <strong>
        1
      </strong>
    </div>
    <div>
      postCharactersMedian: 
      <strong>
        228
      </strong>
    </div>
    <ul>
      postsPerMonth: 
      <li>
        <div>
          January
          : 
          <strong>
            1
          </strong>
        </div>
      </li>
    </ul>
    <div>
      longestPost: 
      mess money hate conductor company duck indulge fly seller neighborhood rush message flush empire upset due climate opposition virgin shy dimension tolerate horror wagon run accumulation huge complex mother resource core era cash
    </div>
  </li>
  <li>
    <strong>
      user_0
    </strong>
    <div>
      postsTotalCount: 
      <strong>
        1
      </strong>
    </div>
    <div>
      postCharactersMedian: 
      <strong>
        307
      </strong>
    </div>
    <ul>
      postsPerMonth: 
      <li>
        <div>
          January
          : 
          <strong>
            1
          </strong>
        </div>
      </li>
    </ul>
    <div>
      longestPost: 
      leaflet execute systematic corn avenue porter attention kit horseshoe trench constitution canvas representative fill nest racism occasion crystal resign lake umbrella revolution tile snack cottage carve mess accountant tired faithful aluminium competition follow variant button basket advice withdrawal bury
    </div>
  </li>
</ul>
`;
